services/mailer/mail_issue.go

```diff
@@ -24,83 +22,83 @@ func fallbackMailSubject(issue *models.Issue) string {
 // 2. Users who are not in 1. but get mentioned in current issue/comment.
 func mailIssueCommentToParticipants(issue *models.Issue, doer *models.User, actionType models.ActionType, content string, comment *models.Comment, mentions []string) error {
 
-	watchers, err := models.GetWatchers(issue.RepoID)
+	// =========== Repo watchers ===========
+	// *Watch
+	rwatchers, err := models.GetWatchers(issue.RepoID)
 	if err != nil {
-		return fmt.Errorf("getWatchers [repo_id: %d]: %v", issue.RepoID, err)
+		return fmt.Errorf("GetWatchers(%d): %v", issue.RepoID, err)
 	}
-	participants, err := models.GetParticipantsByIssueID(issue.ID)
+	watcherids := make([]int64, len(rwatchers))
+	for i := range rwatchers {
+		watcherids[i] = rwatchers[i].UserID
+	}
+	watchers, err := models.GetUsersByIDs(watcherids)
```

Comment:
Consider there are over 1000 people are watching this repository.

Reply:
@lunny I know what you mean, but IMHO it's better to perform fewer database queries (2 vs 1,001) than save 200KB or RAM at most. Are you sure you want to read the users one by one? Otherwise, we need batches, as I've said in the last part of my comment:

>
> This function can be further improved by batching most of the queries (e.g. bring user IDs and retrieve them from user all at once), but that would involve editing several files at models and a harder to review PR.

Gosh, I miss my xrjs operators. 😄

Reply:
@lunny, I'll see if there's a neat way of batching that part at least.